numpy/core/src/umath/ufunc_object.c

```diff
@@ -4396,68 +4406,307 @@ PyUFunc_GenericReduction(PyUFuncObject *ufunc,
-        PyArray_Descr *dtype = NULL;
-        if (!PyArray_DescrConverter2(dtype_obj, &dtype)) {
-            return -1;
-        }
-        if (dtype == NULL) {
-            /* dtype=None, no need to set typetup. */
+        if (dtype_obj == Py_None) {
+            /* If `dtype=None` is passed, no need to do anything */
+            assert(*out_typetup == NULL);
             return 0;
         }
-        *out_typetup = PyTuple_Pack(1, (PyObject *)dtype);
-        Py_DECREF(dtype);
-        if (*out_typetup == NULL) {
+        if (nout == 0) {
+            /* This may be allowed (NumPy does not do this)? */
+            PyErr_SetString(PyExc_TypeError,
+                    "Cannot provide `dtype` when a ufunc has no outputs");
             return -1;
         }
+        signature[nin] = _get_dtype(dtype_obj);
+        if (signature[nin] == NULL) {
+            return -1;
+        }
+        res = _make_new_typetup(nop, signature, out_typetup);
+        goto finish;
     }
-    /* sig and signature are not converted here right now. */
-    return 0;
+
+    assert(signature_obj != NULL);
+    /* Fill in specified_types from the tuple or string (signature_obj) */
+    if (PyTuple_Check(signature_obj)) {
+        int nonecount = 0;
+        Py_ssize_t n = PyTuple_GET_SIZE(signature_obj);
+        if (n == 1 && nop != 1) {
+            /*
+             * Special handling, because we deprecate this path.  The path
+             * probably mainly existed since the `dtype=obj` was passed through
+             * as `(obj,)` and parsed later.
+             */
+            if (PyTuple_GET_ITEM(signature_obj, 0) == Py_None) {
+                PyErr_SetString(PyExc_TypeError,
+                        "a single item type tuple cannot contain None.");
+                goto finish;
+            }
+            if (DEPRECATE("The use of a length 1 tuple for the ufunc "
+                          "`signature` is deprecated. Use `dtype` or  fill the"
+                          "tuple with `None`s.") < 0) {
+                goto finish;
+            }
+            /* Use the same logic as for `dtype=` */
+            res = _get_normalized_typetup(ufunc,
+                    PyTuple_GET_ITEM(signature_obj, 0), NULL, out_typetup);
+            goto finish;
+        }
+        if (n != nop) {
+            PyErr_Format(PyExc_ValueError,
+                    "a type-tuple must be specified of length %d for ufunc '%s'",
+                    nop, ufunc_get_name_cstr(ufunc));
+            goto finish;
+        }
+        for (int i = 0; i < nop; ++i) {
+            PyObject *item = PyTuple_GET_ITEM(signature_obj, i);
+            if (item == Py_None) {
+                ++nonecount;
+            }
+            else {
+                signature[i] = _get_dtype(item);
+                if (signature[i] == NULL) {
+                    goto finish;
+                }
+            }
+        }
+        if (nonecount == n) {
+            PyErr_SetString(PyExc_ValueError,
```

Comment:
Yeah, I agree... The old code deferring parsing to the end had to do this, because by the time it was parsed the normal resolver information would be lost.  But if we parse it up-front (which seems clearly better), there is not really a reason not to support it.